test/validation/api/ml/ml.c

```diff
@@ -0,0 +1,501 @@
+
+	memset(&pool_param, fill, sizeof(pool_param));
+	odp_ml_compl_pool_param_init(&pool_param);
+	CU_ASSERT_EQUAL(pool_param.uarea_size, 0);
+	CU_ASSERT_PTR_NULL(pool_param.uarea_init.args);
+	CU_ASSERT(pool_param.uarea_init.init_fn == NULL);
+	CU_ASSERT(pool_param.cache_size <= global.ml_capa.pool.max_cache_size);
+	CU_ASSERT(pool_param.cache_size >= global.ml_capa.pool.min_cache_size);
+
+	memset(&compl_param, fill, sizeof(compl_param));
+	odp_ml_compl_param_init(&compl_param);
+	CU_ASSERT_PTR_NULL(compl_param.user_ptr);
+
+	memset(&run_param, fill, sizeof(run_param));
+	odp_ml_run_param_init(&run_param);
+	CU_ASSERT_EQUAL(run_param.batch_size, 0);
+	CU_ASSERT_PTR_NULL(run_param.result);
+}
+
+static void test_ml_param_init(void)
+{
+	test_ml_param(0x00);
+	test_ml_param(0xff);
+}
+
+static void test_ml_debug(void)
+{
+	odp_ml_print();
+}
+
+static void ml_compl_pool_create_max_pools(void)
+{
+	int ret;
+	uint32_t i, j;
+	odp_ml_compl_pool_param_t ml_pool_param;
+	uint32_t max_pools = global.ml_capa.pool.max_pools - 1;
+	odp_pool_t compl_pools[max_pools];
+
+	odp_ml_compl_pool_param_init(&ml_pool_param);
+	ml_pool_param.num = global.num_compl;
+	for (i = 0; i < max_pools; i++) {
+		compl_pools[i] = odp_ml_compl_pool_create(NULL, &ml_pool_param);
+
+		if (compl_pools[i] == ODP_POOL_INVALID)
+			break;
+	}
+
+	CU_ASSERT_EQUAL(i, max_pools);
+
+	/* Destroy the created valid pools */
+	for (j = 0; j < i; j++) {
+		ret = odp_pool_destroy(compl_pools[j]);
+		CU_ASSERT_EQUAL(ret, 0);
+
+		if (ret == -1)
+			ODPH_ERR("ML completion pool destroy failed: %u / %u\n", j, i);
+	}
+}
+
+static void compl_pool_info(void)
+{
+	odp_pool_t pool;
+	odp_pool_t compl_pool;
+	odp_pool_info_t pool_info;
+	odp_ml_compl_pool_param_t pool_param;
+
+	/* Create an ML job completion pool */
+	odp_ml_compl_pool_param_init(&pool_param);
+	pool_param.num = global.num_compl;
+
+	compl_pool = odp_ml_compl_pool_create(COMPL_POOL_NAME, &pool_param);
+	CU_ASSERT_NOT_EQUAL_FATAL(compl_pool, ODP_POOL_INVALID);
+
+	/* Verify info about the created ML completion pool compl_pool */
+	pool = odp_pool_lookup(COMPL_POOL_NAME);
+	CU_ASSERT_EQUAL(pool, compl_pool);
+
+	memset(&pool_info, 0x66, sizeof(odp_pool_info_t));
+	CU_ASSERT_EQUAL_FATAL(odp_pool_info(compl_pool, &pool_info), 0);
+
+	CU_ASSERT_STRING_EQUAL(pool_info.name, COMPL_POOL_NAME);
+	CU_ASSERT_EQUAL(pool_info.pool_ext, 0);
+	CU_ASSERT_EQUAL(pool_info.type, ODP_POOL_ML_COMPL);
+	CU_ASSERT_EQUAL(pool_info.ml_pool_param.num, NUM_COMPL);
+	CU_ASSERT_EQUAL(pool_info.ml_pool_param.uarea_size, 0);
+	CU_ASSERT_EQUAL(pool_info.ml_pool_param.cache_size, pool_param.cache_size);
+
+	CU_ASSERT_EQUAL_FATAL(odp_pool_destroy(compl_pool), 0);
+}
+
+static void compl_alloc_max(void)
+{
+	uint64_t u64;
+	odp_event_t event;
+	odp_pool_t compl_pool;
+	odp_ml_compl_pool_param_t pool_param;
+	const int num = global.ml_capa.pool.max_num;
+	odp_ml_compl_t *compl = malloc(num * sizeof(odp_ml_compl_t));
```

Comment:
Variable 'num' should be uint32_t as capa max_num is uint32_t.

A very large capa would lead to a very large malloc. Maybe have max define (1M events) to cap the test memory usage and test run time. Also, should check if malloc failed.

Reply:
Limited to one million in v3, so no need to change the type.